File: src/main/java/jokatu/components/config/WebSocketConfiguration.java
package jokatu.components.config;

import jokatu.util.Json;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.format.support.DefaultFormattingConversionService;
import org.springframework.messaging.simp.annotation.support.SimpAnnotationMethodMessageHandler;
import org.springframework.session.MapSessionRepository;
import org.springframework.session.web.socket.config.annotation.AbstractSessionWebSocketMessageBrokerConfigurer;
import org.springframework.web.socket.config.annotation.EnableWebSocket;
import org.springframework.web.socket.config.annotation.EnableWebSocketMessageBroker;
import org.springframework.web.socket.config.annotation.StompEndpointRegistry;

import javax.annotation.PostConstruct;

/**
 * Set up STOMP web sockets and configure the conversion service for messages to use the default Jackson converter.
 */
@Configuration
@EnableWebSocket
// todo: secure web sockets
@EnableWebSocketMessageBroker
public class WebSocketConfiguration
		extends AbstractSessionWebSocketMessageBrokerConfigurer {

	@Autowired
	private SimpAnnotationMethodMessageHandler simpAnnotationMethodMessageHandler;

	@PostConstruct
	public void init() {
		// How about you use the ConversionService that already exists, rather than creating your own, and then this
		// wouldn't be necessary?  Also how about you update your DestinationVariableMethodArgumentResolver's
		// ConversionService when yours gets set?
		DefaultFormattingConversionService conversionService
				= (DefaultFormattingConversionService) simpAnnotationMethodMessageHandler.getConversionService();
		conversionService.addConverterFactory(new Json.JacksonConverterFactory());
	}

	@Override
	protected void configureStompEndpoints(StompEndpointRegistry registry) {
		registry.addEndpoint("/ws");
	}

	@Bean
	public MapSessionRepository sessionRepository() {
		return new MapSessionRepository();
	}

//	@Override
//	protected void configureInbound(MessageSecurityMetadataSourceRegistry messages) {
//		messages.anyMessage().authenticated();
//	}
}

File: src/main/kotlin/jokatu/components/config/WebSocketConfiguration.kt
package jokatu.components.config

import org.slf4j.getLogger
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.messaging.simp.config.MessageBrokerRegistry
import org.springframework.scheduling.annotation.EnableScheduling
import org.springframework.session.MapSession
import org.springframework.session.MapSessionRepository
import org.springframework.session.web.socket.config.annotation.AbstractSessionWebSocketMessageBrokerConfigurer
import org.springframework.util.AntPathMatcher
import org.springframework.web.socket.config.annotation.EnableWebSocket
import org.springframework.web.socket.config.annotation.EnableWebSocketMessageBroker
import org.springframework.web.socket.config.annotation.StompEndpointRegistry
import org.springframework.web.socket.config.annotation.WebSocketMessageBrokerConfigurer
import java.util.concurrent.ConcurrentHashMap

/**
 * Set up STOMP web sockets.
 */
@Configuration
@EnableScheduling
// todo: secure web sockets
@EnableWebSocketMessageBroker
class WebSocketConfiguration : AbstractSessionWebSocketMessageBrokerConfigurer<MapSession>() {

	override fun configureStompEndpoints(registry: StompEndpointRegistry) {
		registry.addEndpoint("/ws")

		log.debug("{} configured STOMP endpoints", WebSocketConfiguration::class.simpleName)
	}

	override fun configureMessageBroker(registry: MessageBrokerRegistry) {
		registry.enableStompBrokerRelay("/topic/", "/queue/")
		registry.setPathMatcher(AntPathMatcher("."))

		log.debug("{} configured message broker", WebSocketConfiguration::class.simpleName)
	}

	companion object {
		private val log = getLogger(WebSocketConfiguration::class)
	}

	//	@Override
	//	protected void configureInbound(MessageSecurityMetadataSourceRegistry messages) {
	//		messages.anyMessage().authenticated();
	//	}
}
